tests/ui/self/arbitrary_self_types_pin_getref_feature.rs

```diff
@@ -0,0 +1,21 @@
+// Confirms that Pin::get_ref can no longer shadow methods in pointees
+// once arbitrary_self_types is enabled.
+
+#![feature(arbitrary_self_types)]
+#![allow(dead_code)]
+
+use std::pin::Pin;
+use std::pin::pin;
+
+struct A;
+
+impl A {
+    fn get_ref(self: &Pin<&A>) {}  // note &Pin
+}
+
+fn main() {
+    let pinned_a: Pin<&mut A> = pin!(A);
+    let pinned_a: Pin<&A> = pinned_a.as_ref();
+    let _ = pinned_a.get_ref();
+    //~^ ERROR: multiple applicable items
```

Comment:
I don't want to hold up this PR because adding the feature gate check is absolutely necessary as the other test case shows, but what is the plan for stabilizing the feature with regards to this check?

We can't just report this error when the feature is stabilized because that could break existing code. If we need to log an unresolved question for this, that's fine!

Reply:
Good question.

Existing code which could break must fit these criteria:
* It uses an existing standard library smart pointer type - `Box`, `Pin`, `Rc` or `Arc`.
* The smart pointer type must have an inherent method (not just an associated function)
* Case A:
  * That standard library method must take `self` by value
  * The referent must have an identically named method which takes `self: &SmartPointerType<Self>` or `self: &mut SmartPointerType<Self>` 
* Case B:
  * That standard library method must take `&self`
  * The referent must have an identically named method which takes `self: &mut SmartPointerType<Self>`

Standard library methods which might cause these problems:
* Case B: `Pin::as_ref`
* Case A: `Pin::map_unchecked`
* Case A: `Pin::get_ref`
* Case A: `Pin::into_ref`
* Case A: `Pin::get_unchecked_mut`
* Case A: `Pin::map_unchecked_mut`
* There are no cases in `Rc`, `Box` or `Arc`

So folks will get broken if, and only if, they have types with any of these methods:
```rust
fn as_ref(self: &Pin<Self>)
fn as_ref(self: &mut Pin<Self>)
fn map_unchecked(self: &mut Pin<Self>)
fn get_ref(self: &mut Pin<Self>)
fn into_ref(self: &mut Pin<Self>)
fn get_unchecked_mut(self: &mut Pin<Self>)
fn map_unchecked_mut(self: &mut Pin<Self>)
```

But here's the thing. As of now, I don't believe there would be any way to _call_ those methods. Because [method resolution would always pick the method on `Pin` rather than the inner method.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d71b1f74ae5e59417e875288b9e1ba2c) You _could_ call these methods through UFCS, but _in practice_ I don't think anyone will have added these methods to any types, because aside from calling them using UFCS, it's completely pointless to add these methods.

Therefore - I think in practice we won't break anyone.

But, I did want us to consider this fully and decide if it's OK to take this theoretical risk during stabilization instead of now, hence raising this PR.

Reply:
I think it's worthwhile to note this in the stabilization report as something that will begin to be triggered once this is stabilized, though. I'll link it.